[Issue #29] (Team 2) Dynamic Programming Algorithm for Query Rewriter

shiladityasen opened the pull request:
- Added English dictionary file, with frequencies.
- Modified FuzzyParser to implement the dynamic programming algorithm, implemented by @JavierJia which is referenced in the comment thread, under #29
- Modified the QueryRewriterTest to reflect the new expected results from the QueryRewriter

@chenlica : Can you give this a review?

chenlica commented on textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/queryrewriter/FuzzyTokenizer.java:
```diff
@@ -16,12 +18,11 @@
  */
-public class FuzzyTokenizer
-{
+public class FuzzyTokenizer {
```
I think the name `FuzzyTokenizer` is not very accurate.  Can we rename it to `QueryRewriter`?

chenlica replied:
I just noticed that it's a helper class for `QueryRewriter`.  Still I prefer a different name.  Any suggestions?

chenlica replied:
How about `QuerySegmenter`? http://www2003.org/cdrom/papers/poster/p052/xhtml/querysegmentation.html

shiladityasen replied:
Understood... Refactoring to `QuerySegmenter`  
Just asking, after the refactoring, is it ok to have the method named `getFuzzyTokens` ?

chenlica replied:
We want to avoid overloading the word "fuzzy".  So I prefer to use `getTokens()`.

shiladityasen replied:
Understood.

shiladityasen commented:
Actually `QueryRewriter` is the main class... It in turn calls the `FuzzyTokenizer` class to detect missing white spaces. The reason we have separated the two classes is because later, QueryRewriter would also perform "synonym" matching as well.

chenlica commented on textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/queryrewriter/FuzzyTokenizer.java:
```diff
@@ -43,36 +44,99 @@ public FuzzyTokenizer(String phrase) {
+     * When called returns a single-member list containing most-likely tokenization of phrase
+     * Splits phrase into terms by whitespace delimiter
+     * For each term, rewrites(tokenizes) term to detect likely missing spaces
+     * Concatenates all rewritten terms with singlespace delimiter
```
Add an example to the comments.

chenlica commented on textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/queryrewriter/FuzzyTokenizer.java:
```diff
@@ -43,36 +44,99 @@ public FuzzyTokenizer(String phrase) {
+     * Performs most likely tokenization of a single term
+     * Uses Dynamic Programming algorithm approach used for tokenization of Chinese strings
+     * Source - Chinese Tokenizer in SRCH2 codebase
+     *           https://github.com/SRCH2/srch2-ngn/blob/master/src/core/analyzer/ChineseTokenizer.cpp#L197
```
@JavierJia : can you review this code of dynamic programming?

JavierJia replied:
Looks good to me.

shiladityasen commented:
@chenlica : Added an example in the comments as suggested... Kindly take a look

chenlica commented:
In the `QueryRewriter` class, is it better to make `fuzzyTokenizer.getFuzzyTokens()` a static function if it only needs a query string?

chenlica commented on textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/queryrewriter/FuzzyTokenizer.java:
```diff
@@ -43,36 +44,106 @@ public FuzzyTokenizer(String phrase) {
     }
 
     /**
+     * When called returns a single-member list containing most-likely tokenization of phrase
```
For line 42 `public FuzzyTokenizer(String phrase`, rename `phrase` to `queryString` since `phrase` has a special meaning in other modules.

shiladityasen replied:
It does make sense for `fuzzyTokenizer.getFuzzyTokens()` to be a static function. We are making changes as suggested.

chenlica commented on textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/queryrewriter/FuzzyTokenizer.java:
```diff
@@ -43,36 +44,106 @@ public FuzzyTokenizer(String phrase) {
+        String[] terms = phrase.split("\\s+");
+        String rewrittenPhrase = "";
+        for(String term : terms)
+            rewrittenPhrase += " " + rewriteTerm(term);
```
It means the new query will always start with " ".  Is that what we want?  I think it should start with a word, then followed by a space.

shiladityasen replied:
@chenlica : I am doing this to ensure the concatenation of rewritten terms are space separated.  
To ensure the new query does not start with a " ", I am returning the substring starting from index 1.

chenlica replied:
I understood it.  Let's keep it then.

chenlica commented on textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/queryrewriter/FuzzyTokenizer.java:
```diff
@@ -43,36 +44,106 @@ public FuzzyTokenizer(String phrase) {
+        String rewrittenPhrase = "";
+        for(String term : terms)
+            rewrittenPhrase += " " + rewriteTerm(term);
+        rewrittenPhrase = rewrittenPhrase.substring(1);
```
I think this statement is getting rid of the first space.  If so, can we rewrite the code without this extra step?

chenlica commented on textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/queryrewriter/FuzzyTokenizer.java:
```diff
@@ -43,36 +44,106 @@ public FuzzyTokenizer(String phrase) {
      * @return
      */
-    public List<String> getFuzzyTokens() {
+    public List<String> bruteGetFuzzyTokens() {
```
If the bruce forth method is no longer needed, I suggest we remove the related code completely.

shiladityasen replied:
The DP method only returns the most likely tokenization. The brute force method is the only method till now which returns all possible tokenizations and that's why we were thinking of still keeping it around till we find a substitute for that use-case.

chenlica replied:
OK.

chenlica replied:
Please add a comment to the codebase to indicate that this function is not used, and why we keep it.  Also please keep the test cases for it.

shiladityasen replied:
Added a comment to indicate this function is not currently used and why we are keeping it.  
However, regarding keeping the test cases there might be an issue. The test cases interact with the `QueryRewriter` class by the `getNextTuple` function. There is no way to indicate with tokenizer `getNextTuple` should call (i.e. DP or Brute Force). Since the signature for the `IOperator` does not allow for such an option in `getNextTuple`, the tests cannot be performed on both the tokenizers.  
@chenlica : Kindly suggest what we should do regarding the tests.

chenlica replied:
One way to keep the code is to pass a special parameter to the QueryRewriter constructor to indicate whether it's a brute forth method.  By default, it's not.  Then the test cases can still be kept by passing this parameter.  Does it make sense?  @shiladityasen

shiladityasen replied:
@chenlica : Yes Professor, making the changes as suggested.

chenlica commented:
High-level comments:

- Make FuzzyTokenizer.getFuzzyTokens() a static function.
- Possibly rename FuzzyTokenizer to QuerySegmenter.  If you don't agree, that's fine.
- Wait for @JavierJia to review the DP algorithm.

chenlica commented on textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/queryrewriter/FuzzyTokenizer.java:
```diff
@@ -43,36 +44,106 @@ public FuzzyTokenizer(String phrase) {
+            }
+        }
+
+        String tokenization = "";
```
Rename `tokenization` to `tokenizedQuery`.

chenlica commented on textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/queryrewriter/FuzzyTokenizer.java:
```diff
@@ -43,36 +44,106 @@ public FuzzyTokenizer(String phrase) {
+        String tokenization = "";
+
+        for(int pos = size-1; pos > 0; pos = preBestGap[pos]-1) {
+            tokenization = " " + term.substring(preBestGap[pos], pos+1) + tokenization;
```
Change it to 
```
  if (pos > 1)
            tokenization = " " + term.substring(preBestGap[pos], pos+1) + tokenization;
 else
            tokenization = term.substring(preBestGap[pos], pos+1) + tokenization;
```

Then we can just return `tokenization`.

shiladityasen replied:
In this phase, we reverse-engineer the tokenization by jumping over the preBestGap indices.  
Changing `pos > 0` to `pos > 1` might not help since `1` might not be the index of a likely whitespace. `0` definitely works since the word starts from that index.

chenlica replied:
OK.

shiladityasen commented:
- Refactored FuzzyTokenizer to QuerySegmenter
- Refactored `phrase` to `queryString`
- Refactored `getFuzzyTokens` to `getTokens`
- Converted `getTokens` to static function and made corresponding changes to `QueryRewriter`

@chenlica : Kindly take a look.

chenlica commented:
To unblock you, no need to wait for the review from @JavierJia .  Go ahead to do the merge after making those changes.  We will let @JavierJia doublecheck the logic even after the merge.

chenlica commented:
Looks good.  Please go ahead to do the merge!

shiladityasen commented:
Missed refactoring a variable name...

chenlica commented:
Looks good.  Please go ahead to do the merge.

shiladityasen commented:
- Overloaded `QueryRewriter` constructor to accept flag for returning all possible tokenizations
- Included corresponding test cases in `QueryRewriterTest`
- Cleaned some noisy entries in word-frequency map

@chenlica : Can you take a look at the changes?

chenlica commented on textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/queryrewriter/QueryRewriter.java:
```diff
@@ -64,7 +79,11 @@ public ITuple getNextTuple() throws Exception {
-            List<String> queryStrings = QuerySegmenter.getTokens(searchQuery);
+            List<String> queryStrings;
+            if(allSegmentations)
+                queryStrings = QuerySegmenter.bruteGetTokens(searchQuery);
```
`bruteGetTokens` -> `getTokensBruteForth`

shiladityasen replied:
We were thinking of refactoring `bruteGetTokens` -> `getAllTokens` and `getTokens` -> `getLikelyTokens`.  
`getAllTokens` can internally call `rewriteTermBruteForce` and `getLikelyTokens` can call `rewriteTermDP`  
@chenlica : Do you think that would be ok?

chenlica replied:
Sounds good.  The name `rewriteTermDP` can be better.

chenlica commented on textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/queryrewriter/QueryRewriter.java:
```diff
@@ -31,15 +32,26 @@
     private ITuple itupleResult;
```
`itupleResult` -> `sourceTuple`?

chenlica commented:
I left some minor comments on your commit.  After taking care of them, please do the merge.  Thanks!

shiladityasen commented:
@chenlica : Refactored changes as suggested. I am going ahead with the merge.  
Thank you, Professor.
